python-eval-mathrepl/mathrepl.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+
+# File: mathrepl.py
+
+"""MathREPL, a math expression evaluator using Python's eval() and math."""
+
+import math
+
+__version__ = "1.0"
+__author__ = "Leodanis Pozo Ramos"
+
+ALLOWED_NAMES = {
+    name: obj for name, obj in math.__dict__.items() if "__" not in name
+}
+
+PS1 = "mr>>"
+
+WELCOME = f"""
+MathREPL {__version__}, your Python math expressions evaluator!
+Enter a valid math expression after the prompt "{PS1}".
+Type "help" for more information.
+Type "quit" or "exit" to exit.
+"""
+
+USAGE = f"""
+Usage:
+Build math expressions using numeric values and operators.
+Use any of the following functions and constants:
+
+{', '.join(ALLOWED_NAMES)}
+"""
+
+
+def evaluate(expression: str) -> float:
+    """Evaluate a math expression."""
+    # Compile and validate syntax
+    try:
+        code = compile(expression, "<string>", "eval")
+    except SyntaxError:
+        raise SyntaxError("Invalid input expression")
+
+    # Validate allowed names
+    for name in code.co_names:
+        if name not in ALLOWED_NAMES:
+            raise NameError(f'The use of "{name}" is not allowed')
+
+    return eval(code, {"__builtins__": {}}, ALLOWED_NAMES)
+
+
+def main() -> None:
+    """Main loop: Read and evaluate user input."""
+    print(WELCOME)
+    while True:
+        # Read user's input
+        try:
+            expression = input(f"{PS1} ")
+        except (KeyboardInterrupt, EOFError):
+            raise SystemExit()
+
+        # Handle special commands
+        if expression.lower() == "help":
+            print(USAGE)
+            continue
+        if expression.lower() in {"quit", "exit"}:
+            raise SystemExit()
+
+        # Evaluate the expression
+        try:
+            result = evaluate(expression)
+        except Exception as err:
+            print(err)
```

Comment:
Ah, I didn't catch this one on the article.  You should add `continue` below the `print(err)` here, to avoid printing the previous result (which may be undefined on the first loop iteration)

Reply:
My mistake, I'll fix this.

Reply:
@lpozo This isn't fixed here

Reply:
I moved the print call to the `try` block, I think this fixes the problem, don't you?

Reply:
This is where it was originally. The problem is that if there is an exception, the error is printed and then the script runs the next line, which is to print the result. If the result is undefined, it results in another, unhandled, error. If the result is defined from a previous successful calculation, the error message is printed and then the success message is printed.

Reply:
Yes, but that code isn't there anymore. We don't have any statement below `print(err)` in this new version. Do you think that we still have the same problem?

Reply:
Oh, I see you moved the other print call inside the `try` block. Well, it definitely won't have the same problem, but I'm a little torn about that. Ideally, you'd do as little as possible inside the `try` statement, to make sure that you only raised errors you wanted to handle. On the other hand, you're deliberately catching all errors in this block on purpose...

After some thought, I think it would be better to have the `print("The result is...")` outside the `try` block. The reason is that if `print()` encounters an error for some reason, you don't want to be catching that error, the user should definitely know if `print()` is having a problem, and it should probably "crash" the program. In addition, having a separate print call for the result is somehow **clearer** to me, it is more obvious what is being done in each step for the reader. Does that all make sense? If you strongly disagree, feel free to leave the printing of the result in the `try` block.